go/teams/transactions.go

```diff
@@ -562,6 +555,29 @@ func (tx *AddMemberTx) AddMemberByAssertionOrEmail(ctx context.Context, assertio
 		return "", uv, false, err
 	}
 	return "", uv, true, nil
+
+}
+
+// AddOrInviteMemberByAssertionOrEmail adds an assertion to the team. It can handle
+// three major cases:
+//  1. joe OR joe+foo@reddit WHERE joe is already a keybase user, or the assertions map to a unique keybase user
+//  2. joe@reddit WHERE joe isn't a keybase user, and this is a social invitations
+//  3. [bob@gmail.com]@email WHERE server-trust resolution is performed and either TOFU invite is created
+//     or resolved member is added. Same works with `@phone`.
+// **Does** attempt to resolve the assertion, to distinguish between case (1), case (2) and an error
+// The return values (uv, username) can both be zero-valued if the assertion is not a keybase user.
+func (tx *AddMemberTx) AddOrInviteMemberByAssertionOrEmail(ctx context.Context, assertion string, role keybase1.TeamRole, botSettings *keybase1.TeamBotSettings) (
```

Comment:
maybe it's time to redesign this API. It's only used internally in the client, so we don't need to keep it compatible, and now we know more about how we want to use this stuff than back then when it was being written.

Reply:
A general method like: `tx.addByAssertion(assertion, role)` - that figures out whether to do `change_membership`, `invite`, sweep old versions etc. seems like what most callers need, but there is a bunch that want only specific things to happen, e.g. open team access requests only tries to add cryptomembers, or there is the entire implicit team ruleset.